File: answers/leetcode/Merge k Sorted Lists/Merge k Sorted Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """

        heap = []
        for i in lists:
            if i:
                heap.append((i.val, i))
        heapq.heapify(heap)
        dummy = ListNode(-1)
        current = dummy
        while len(heap):
            pop = heapq.heappop(heap)
            current.next = pop[1]
            if pop[1].next:
                heapq.heappush(heap, (pop[1].next.val, pop[1].next))
            current = current.next
        return dummy.next

File: answers/leetcode/Merge k Sorted Lists/Merge k Sorted Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) {
            return null;
        }
        PriorityQueue<ListNode> queue = new PriorityQueue<>(lists.length, (a, b) -> a.val - b.val);
        for (ListNode list : lists) {
            if (list != null) {
                queue.add(list);
            }
        }
        ListNode head = new ListNode(-1);
        ListNode current = head;
        while (!queue.isEmpty()) {
            ListNode next = queue.remove();
            if (next.next != null) {
                queue.add(next.next);
            }
            current.next = next;
            current = current.next;
        }
        return head.next;
    }
}